File: MatrixDiagonal.py
class Solution:
    """
    Time complexity - O(n*m). n - no of rows in matrix m - no of columns in matrix
    Space complexity - O(1)
    """
    def findDiagonalOrder(self, mat: list[list[int]]) -> list[int]:
        n = len(mat)
        m = len(mat[0])
        row = 0
        col = 0
        result = []
        up = True
        for x in range(0, n*m):
            # upwards diagonal
            if up:
                while row >=0 and col < m:
                    result.append(mat[row][col])
                    row -= 1
                    col += 1
                # set the row and col for next traversal
                if row < 0 and col < m:
                    row = 0
                else:
                    row += 2
                    col -= 1
                up = False
            # downwards diagonal
            else:
                while row < n and col >= 0:
                    result.append(mat[row][col])
                    row += 1
                    col -= 1
                # set the row and col for next traversal
                if row < n and col < 0:  
                    col = 0
                else:
                    row -= 1
                    col += 2
                up = True
        return result
                    
                
                
        

File: matrix_diagonal.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        if(mat == null || mat.length==0)
        {
            return new int[0];
        }
        
        int direction = 1;
        
        int m = mat.length;
        int n = mat[0].length;
        
        int[] result = new int[m*n];
        int i = 0;
        
        int row = 0;
        int col = 0;
        
        while(row != m && col!= n)
        {
            result[i] = mat[row][col];
            
            //if direction up
            if(direction == 1)
            {
                if(col==n-1)
                {
                    row++;
                    direction = -1;
                }
                else if(row ==0)
                {
                    col++;
                    direction = -1;
                }
                else{
                    col++;
                    row--;
                }
            }
            else{
                if(row == m-1)
                {
                    col++;
                    direction = 1;
                }
                else if(col == 0)
                {
                    row++;
                    direction = 1;
                }
                else{
                    row++;
                    col--;
                }
            }
            i++;
        }
        return result;
        
    }
}

//runtime: O(mn)
//space: O(1)
